MetaMask/metamask-extension
# Fix disabled 'eth_sign' event & add 'Request Disabled' event type

## Explanation

Fixes https://github.com/MetaMask/metamask-extension/issues/17974

- Fixes bug where 'Signature Approved' event was incorrectly called when 'eth_sign' rpc method is called while the method is disabled by advanced settings
- Replaces 'Signature Approved' event with new 'Signature Failed' event when the disabled 'eth_sign' event is called (spec discussed [here in Slack](https://consensys.slack.com/archives/G01FXS64DFA/p1678102135498189))

**note:** 
1. sometimes, the 'Signature Failed' effect may be received before 'Signature Request', but the 'sent_at' time still records it being sent earlier. 
2. clean-up code, including more use of constants, is applied to this follow-up PR https://github.com/MetaMask/metamask-extension/pull/18008



## Screenshots/Screencaps



### Before


### After


<img width="1473" alt="Screen Shot 2023-03-13 at 12 10 39 PM" src="https://user-images.githubusercontent.com/20778143/224849236-ea8277f5-ebbb-4dd4-a63c-80210f13e929.png">
<img width="1470" alt="Screen Shot 2023-03-13 at 12 10 45 PM" src="https://user-images.githubusercontent.com/20778143/224849240-1e598021-54ab-4db9-9311-193aa8712644.png">


## Manual Testing Steps

1. Open Segment or open background console for check the network requests to Segment
2. Disable the functionality of eth_sign from MetaMask Advanced Settings
3. On another window, go to test dapp
4. Connect to test dapp
5. Click Eth Sign
6. See Signature Request and Request Denied events are fired. Signature Approval should no longer be fired.




## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
